mautic/mautic
# Lead import improvement

### New features
- Fields are pre-selected on import fields matching form if the field name in CSV has exact name.
- Added fields for importing these lead properties: "firstname", "lastname", "email", "dateAdded", "createdByUser", "dateModified", "modifiedByUser", "lastActive", "dateIdentified", "ip", "points".
- Added option to add imported leads to a Lead List.

### How to test
For testing new features you'll need at least 2 Mautic users. You can use this CSV and modify values in it as you'll try to test stuff:
```
"firstname", "lastname", "email", "dateAdded", "createdByUser", "dateModified", "modifiedByUser", "lastActive", "dateIdentified", "ip", "points", "doNotEmail"
"John", "Imported 1", "imported1@gmail.com", "dd", "admin@email.com", "2010-10-10 10:00:00", "user2", "2014-10-01 10:00:00", "2014-10-01 11:00:00", "10.0.0.36, 98.34.219.3", "2", "1"
"John", "Imported 2", "imported2@gmail.com", "2014-11-12 10:00:00", "admin", "2014-10-10 10:00:00", "user2", "2014-10-01 10:00:00", "2014-10-01 11:00:00", "10.0.0.36", "2", "0"
"John", "Imported 3", "imported3@gmail.com", "2014-11-12 10:00:00", "user2", "2014-10-10 10:00:00", "user2", "2014-10-01 10:00:00", "2014-10-01 11:00:00", "10.0.0.36", "2", "true"
"John", "Imported 4", "imported4@gmail.com", "2014-11-12 10:00:00", "user2@gmail.com", "2014-10-10 10:00:00", "user2", "2014-10-01 10:00:00", "2014-10-01 11:00:00", "10.0.0.36", "2", "false"
"John", "Imported 5", "imported5@gmail.com", "2014-11-12 10:00:00", "4", "2014-10-10 10:00:00", "user2", "2014-10-01 10:00:00", "2014-10-01 11:00:00", "10.0.0.36", "1000000", "yup"
"", "", "imported6@gmail.com", "", "", "", "", "", "", "", "", ""
```
Try to import weird values like people sometimes do and check if and what was imported. It should never throw an error during import and always deal with that value the way it makes sense.

#### Special behavior of fields:
- `createdByUser`, `modifiedByUser`, `dateModified`, `dateAdded` and `modifiedByUser`: These should be imported only for new leads. For existing leads it should be set to current user and current time (or the user and time who and when the lead was saved).
- `ip`: can have more IP addresses separated by comma
- `points`: Will be imported for new leads because otherwise we don't know if we should add points for existing lead to the points it already has or reset his point count to that amount.
- `doNotEmail`: can be "true" or "1" to lead to be added to ignore list. Other values should be ignored.

Fields should be pre-selected with testing CSV (above). If you change column name, you'll have to select it before hitting Import button. Don't forget to test Lead List option. Check imported values in the database.
